netlogo-core/src/main/fileformat/NLogoXMLLoader.scala

```diff
@@ -0,0 +1,318 @@
+        reader.next match {
+          case XMLStreamConstants.START_ELEMENT =>
+            parseElement(reader, acc.copy(children = acc.children :+ readXMLElement(reader)))
+          case XMLStreamConstants.END_ELEMENT =>
+            acc
+          case XMLStreamConstants.CHARACTERS =>
+            val newAcc = acc.copy(text = new Regex(s"]]${XMLElement.CDATA_ESCAPE}>").replaceAllIn(reader.getText, "]]>"))
+            parseElement(reader, newAcc)
+          case x =>
+            throw new Exception(s"Unexpected value found while parsing XML: ${x}")
+        }
+      else
+        acc
+    }
+
+    val attributes =
+      (0 until reader.getAttributeCount).
+        map((i) => reader.getAttributeLocalName(i) -> reader.getAttributeValue(i)).
+        toMap
+
+    parseElement(reader, XMLElement(reader.getLocalName, attributes,  "", Seq[XMLElement]()))
+
+  }
+
+  private def writeXMLElement(writer: XMLStreamWriter, element: XMLElement): Unit = {
+
+    writer.writeStartElement(element.name)
+
+    for ((key, value) <- element.attributes)
+      writer.writeAttribute(key, value)
+
+    if (element.text.isEmpty)
+      element.children.foreach(writeXMLElement(writer, _))
+    else
+      writeCDataEscaped(writer, element.text)
+
+    writer.writeEndElement()
+
+  }
+
+  private def writeCDataEscaped(writer: XMLStreamWriter, text: String): Unit = {
+    writer.writeCData(new Regex("]]>").replaceAllIn(text, "]]" + XMLElement.CDATA_ESCAPE + ">"))
+  }
+
+  private def isCompatible(extension: String): Boolean =
+    (Version.is3D && extension == "nlogo3d") || (!Version.is3D && extension == "nlogo")
+
+  private def isCompatible(uri: URI): Boolean = {
+    val extension = AbstractModelLoader.getURIExtension(uri)
+    extension.isDefined && isCompatible(extension.get)
+  }
+
+  def readModel(uri: URI): Try[Model] = {
+    val text =
+      if (uri.getScheme == "jar")
+        Source.fromInputStream(uri.toURL.openStream).mkString
+      else
+        Source.fromURI(uri).mkString
+    readModel(text, AbstractModelLoader.getURIExtension(uri).getOrElse(""))
+  }
+
+  def readModel(source: String, extension: String): Try[Model] = {
+
+    if (isCompatible(extension)) {
+
+      val reader = XMLInputFactory.newFactory.createXMLStreamReader(new StringReader(source))
+
+      while (reader.hasNext && reader.next != XMLStreamConstants.START_ELEMENT) {}
+
+      val element = readXMLElement(reader)
+
+      reader.close()
+
+      element.name match {
+        case "model" =>
+
+          import Model.{ defaultCode, defaultShapes, defaultLinkShapes }
+
+          val version = element("version")
+          val snapToGrid = element("snapToGrid", "false").toBoolean
+
+          val settings  = new Section("org.nlogo.modelsection.modelsettings", ModelSettings(snapToGrid))
+
+          val model = Model(defaultCode, List(), defaultInfo, version, defaultShapes, defaultLinkShapes, List(settings), Seq(), Seq())
+
+          element.children.foldLeft(Try(model)) {
```

Comment:
@TheBizzle The way you refactored this makes it fail every time, because you can't instantiate the Model class without a View widget. I will revert this to what I had before so that it works as expected, unless you have another solution.

Reply:
Nevermind, I realized it's probably fine to just instantiate it with a default View widget and assume it will be replaced when the widgets element gets parsed.